src/react-apps/applications/ux-editor/src/utils/arrayHelpers/arrayLogic.ts

```diff
@@ -0,0 +1,110 @@
+import { createSelector } from 'reselect';
+
+export interface IUpdateArrayObj {
+  array: any[];
+  object: any;
+}
+
+export interface ITwoArraysObj {
+  array: any[];
+  order: any [];
+}
+
+const pushOrPop = (obj: IUpdateArrayObj) => {
+  const tempObj = obj.array.find((o: any) => o.id === obj.object.id);
+  if (!tempObj) {
+    obj.array.push(obj.object);
+    obj.array.sort(compareOrderNum);
+  } else {
+    const index = obj.array.indexOf(tempObj);
+    obj.array.splice(index, 1);
+  }
+  if (obj.array.length > 0) {
+    markFirstAndLastObject(obj.array);
+  }
+  const cloneOfObj = JSON.parse(JSON.stringify(obj));
+  return cloneOfObj;
+};
+
+const changeOrderNum = (obj: ITwoArraysObj) => {
+  obj.array.forEach((component: any) => {
+    if (obj.order.indexOf(component.id) >= 0) {
+      component.order = obj.order.indexOf(component.id);
+    }
+  });
+  const cloneOfObj = JSON.parse(JSON.stringify(obj));
+  return cloneOfObj;
+};
+
+const compareOrderNum = (a: any, b: any) => {
+  if (a.order < b.order) {
+    return -1;
+  } else if (a.order > b.order) {
+    return 1;
+  } else if (a.order === b.order) {
+    b.order = a.order + 1;
+  }
+  return 0;
+};
+
+const findMissing = (a: any[]) => {
+  const maximum = Math.max.apply(Math, a.map((o) => {
+    return o.order;
+  }));
+  const minimum = Math.min.apply(Math, a.map((o) => {
+    return o.order;
+  }));
+  let index = 0;
+  Array.from(Array(maximum).keys()).forEach((i: number) => {
```

Comment:
No easier way to write this loop?

Reply:
no 😢

Reply:
After discussion, seems like it's not any easier way to write this loop.